indexing-service/src/test/java/io/druid/indexing/overlord/http/OverlordResourceTest.java

```diff
@@ -0,0 +1,305 @@
+    // Test Overlord resource stuff
+    overlordResource = new OverlordResource(taskMaster, new TaskStorageQueryAdapter(taskStorage), null, null, null);
+    Response response = overlordResource.getLeader();
+    Assert.assertEquals(druidNode.getHostAndPort(), response.getEntity());
+    String taskId_0 = "0";
+    NoopTask task_0 = new NoopTask(taskId_0, 0, 0, null, null);
+    response = overlordResource.taskPost(task_0);
+    Assert.assertEquals(200, response.getStatus());
+    Assert.assertEquals(ImmutableMap.of("task", taskId_0), response.getEntity());
+    // Duplicate task - should fail
+    response = overlordResource.taskPost(task_0);
+    Assert.assertEquals(400, response.getStatus());
+    // Task payload for task_0 should be present in taskStorage
+    response = overlordResource.getTaskPayload(taskId_0);
+    Assert.assertEquals(task_0, ((Map) response.getEntity()).get("payload"));
+    // Task not present in taskStorage - should fail
+    response = overlordResource.getTaskPayload("whatever");
+    Assert.assertEquals(404, response.getStatus());
+    // Task status of the submitted task should be running
+    response = overlordResource.getTaskStatus(taskId_0);
+    Assert.assertEquals(taskId_0, ((Map) response.getEntity()).get("task"));
+    Assert.assertEquals(TaskStatus.running(taskId_0).getStatusCode(), ((TaskStatus)((Map) response.getEntity()).get("status")).getStatusCode());
+    // Simulate completion of task_0
+    taskCountDownLatches[Integer.parseInt(taskId_0)].countDown();
+    // Wait for taskStorage sync
+    Thread.sleep(150);
+    response = overlordResource.getTaskStatus(taskId_0);
+    Assert.assertEquals(200, response.getStatus());
+    Assert.assertEquals(TaskStatus.success(taskId_0).getStatusCode(), ((TaskStatus)((Map) response.getEntity()).get("status")).getStatusCode());
+    // Manually insert task in taskStorage
+    String taskId_1 = "1";
+    NoopTask task_1 = new NoopTask(taskId_1, 0, 0, null, null);
+    taskStorage.insert(task_1, TaskStatus.running(taskId_1));
+    response = overlordResource.getWaitingTasks();
+    // 1 waiting task that was manually inserted
+    Assert.assertEquals(1, (((List) response.getEntity()).size()));
+    // Wait for taskStorage sync
+    Thread.sleep(150);
+    // Simulate completion of task_1
+    taskCountDownLatches[Integer.parseInt(taskId_1)].countDown();
+    // Wait for taskQueue to handle success status of task_1
+    Thread.sleep(50);
+    // should return number of tasks which are not in running state
+    response = overlordResource.getCompleteTasks();
+    Assert.assertEquals(2, (((List) response.getEntity()).size()));
+    taskMaster.stop();
+    Assert.assertFalse(taskMaster.isLeading());
+    EasyMock.verify(taskLockbox, taskActionClientFactory);
+  }
+
+  @After
+  public void tearDown() throws Exception
+  {
+    tearDownServerAndCurator();
+  }
+
+  public static class MockTaskRunner implements TaskRunner{
```

Comment:
It will be hard to mock as getKnownTasks is called multiple times, basically each time manage() method is run. Also if we mock it then we cannot put ListenableFuture for the TaskStatus in the TaskRunnerWorkItem because of which callbacks to handle the status of task will fail. Can you please specify what are the issues in having a mock task runner ?